go/chat/signencrypt/codec.go

```diff
@@ -0,0 +1,530 @@
+// This is a construction for encrypting and signing a message, using a
+// symmetric encryption key and a signing keypair, in a way that supports safe
+// streaming decryption. We need this for chat attachments because we've chosen
+// to use signing keys for authenticity in chat, and we don't want one
+// participant to be able to modify another's attachment, even with an evil
+// server's help. It's *almost* enough that we record the hash of the
+// attachment along with the symmetric key used to encrypt it, but that by
+// itself doesn't allow safe streaming decryption. Instead, we use this
+// construction to sign each chunk of the attachment as we encrypt it. (Note
+// that it's still possible for a sender with the server's help to modify their
+// *own* attachments after the fact, if clients aren't checking the hash. This
+// isn't perfect, but it's better than any participant being able to do it.)
+//
+// This file has 100% test coverage. Please keep it that way :-)
+//
+// Seal inputs:
+// - plaintext bytes (streaming is fine)
+// - a crypto_secretbox symmetric key
+// - a crypto_sign private key
+// - a globally unique (with respect to these keys) 16-byte nonce
```

Comment:
And the hash. As long as we *promise* not to reuse the keys :) Also slight preference on my part for a meaningful constant ASCII string instead of just zeroes.